sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,335 @@ bool WinHttpTransport::VerifyCertificatesInChain(
+   *
+   * Special consideration for the WINHTTP_CALLBACK_STATUS_SENDING_REQUEST - this callback is
+   * called during the TLS connection - if a TLS roto certificate is configured, we verify that the
+   * certificate chain sent from the server contains the certificate the HTTP client was configured
+   * with. If it is, we accept the connection, if it is not, we abort the connection, closing the
+   * incoming request handle.
+   */
+  void WinHttpAction::OnHttpStatusOperation(
+      HINTERNET hInternet,
+      DWORD internetStatus,
+      LPVOID statusInformation,
+      DWORD statusInformationLength)
   {
-    DWORD bufferLength = sizeof(PCCERT_CONTEXT);
-    if (!WinHttpQueryOption(
-            hInternet,
-            WINHTTP_OPTION_SERVER_CERT_CONTEXT,
-            reinterpret_cast<void*>(serverCertificate.addressof()),
-            &bufferLength))
+    Log::Write(
+        Logger::Level::Informational,
+        "Status operation: " + std::to_string(internetStatus) + "("
+            + InternetStatusToString(internetStatus) + ")");
+    if (internetStatus == WINHTTP_CALLBACK_STATUS_SECURE_FAILURE)
+    {
+      Log::Write(Logger::Level::Error, "Security failure. :(");
+    }
+    else if (internetStatus == WINHTTP_CALLBACK_STATUS_REQUEST_ERROR)
+    {
+      WINHTTP_ASYNC_RESULT* asyncResult = static_cast<WINHTTP_ASYNC_RESULT*>(statusInformation);
+      Log::Write(
+          Logger::Level::Error,
+          "Request error. " + GetErrorMessage(asyncResult->dwError) + " "
+              + std::to_string(asyncResult->dwResult));
+      CompleteActionWithError(asyncResult->dwResult, asyncResult->dwError);
+    }
+    else if (internetStatus == WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+    {
+      // We will only set the Status callback if a root certificate has been set. There is no action
+      // which needs to be completed for this notification.
+      m_httpRequest->HandleExpectedTlsRootCertificates(hInternet);
+    }
+    else if (internetStatus == m_expectedStatus)
     {
-      GetErrorAndThrow("Could not retrieve TLS server certificate.");
+      if (internetStatus == WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE)
+      {
+        // A WinHttpSendRequest API call has completed, complete the current action.
+        CompleteAction();
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_WRITE_COMPLETE)
+      {
+        // A WinHttpWriteData call has completed, complete the current action.
+        CompleteAction();
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_HEADERS_AVAILABLE)
+      {
+        // Headers for an HTTP response are available, complete the current action.
+        CompleteAction();
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_READ_COMPLETE)
+      {
+        // A WinHttpReadData call has completed. Complete the current action, including the amount
+        // of data read.
+        CompleteActionWithData(statusInformationLength);
+      }
```

Comment:
nit: Given most of these if-blocks are identical, other than the comment, would a switch be cleaner? Just a thought.

Reply:
I guess? I want to keep each of these broken out separately because the message does not always reflect the API call which is being completed, but it does make the code a bit more compact.